lib/phoenix/live_dashboard/pages/ecto_stats_page.ex

```diff
@@ -3,51 +3,112 @@ defmodule Phoenix.LiveDashboard.EctoStatsPage do
   use Phoenix.LiveDashboard.PageBuilder
   import Phoenix.LiveDashboard.Helpers
 
-  @compile {:no_warn_undefined, [Decimal, EctoPSQLExtras]}
+  @compile {:no_warn_undefined, [Decimal, EctoPSQLExtras, {Ecto.Repo, :all_running, 0}]}
   @disabled_link "https://hexdocs.pm/phoenix_live_dashboard/ecto_stats.html"
+  @page_title "Ecto Stats"
 
   @impl true
-  def init(%{repo: nil}), do: {:ok, %{repo: nil, ecto_options: []}}
+  def init(%{repos: repos, ecto_psql_extras_options: ecto_options}) do
+    capabilities = for repo <- List.wrap(repos), do: {:process, repo}
+    repos = repos || :auto_discover
 
-  def init(%{repo: repo, ecto_psql_extras_options: ecto_options}),
-    do: {:ok, %{repo: repo, ecto_options: ecto_options}, process: repo}
+    {:ok, %{repos: repos, ecto_options: ecto_options}, capabilities}
+  end
 
   @impl true
-  def mount(_params, %{repo: repo, ecto_options: ecto_options}, socket) do
-    {:ok,
-     assign(socket,
-       repo: repo,
-       ecto_options: ecto_options,
-       info_module: info_module_for(repo)
-     )}
+  def mount(_params, %{repos: repos, ecto_options: ecto_options}, socket) do
+    socket = assign(socket, ecto_options: ecto_options)
+
+    fun =
+      case repos do
+        :auto_discover ->
+          fn _ -> auto_discover() end
+
+        repos when is_list(repos) ->
+          &filter_repos_with_extra/1
+      end
+
+    case fun.(repos) do
+      {:ok, repos} ->
+        {:ok, assign(socket, :repos, repos)}
+
+      {:error, error} ->
+        {:ok, assign(socket, :error, error)}
+    end
+  end
+
+  defp auto_discover do
+    with true <- function_exported?(Ecto.Repo, :all_running, 0),
+         [_ | _] = repos <- Ecto.Repo.all_running(),
+         [_ | _] = named_repos <- Enum.filter(repos, &is_atom/1),
+         [_ | _] = repos_with_extra <- Enum.filter(named_repos, &extra_available?/1) do
+      {:ok, repos_with_extra}
+    else
+      false ->
+        {:error, :repos_auto_discovery_not_available}
+
+      [] ->
+        {:error, :no_ecto_repos_available}
+    end
+  end
+
+  defp filter_repos_with_extra(repos) do
+    repos_with_extra = Enum.filter(repos, &extra_available?/1)
+
+    if repos_with_extra == [] do
+      {:error, :no_ecto_repos_available}
+    else
+      {:ok, repos_with_extra}
+    end
   end
 
   @impl true
-  def menu_link(%{repo: nil}, _capabilities) do
+  def menu_link(%{repos: []}, _capabilities) do
```

Comment:
We should disable if it is set to an empty list now, given we have autodiscovery.

I think the workflow is this now:

1. If repos is an empty list, don't show the menu item
2. If autodiscover, shows the available repos. Show an error saying "No Ecto Repo found" if none is found
3. If repos are explicitly given, always list all repos

Reply:
I changed to something in those lines. I'm not sure how to test the case when auto discover doesn't find repos (because the repo is always available). Maybe in a multi node environment I can simulate that easily. Thoughts?